nl_server/lib/nl_embeddings.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Managing the embeddings."""
+from datasets import load_dataset
+from google.cloud import storage
+from sentence_transformers import SentenceTransformer
+from sentence_transformers.util import semantic_search
+from typing import Dict, List, Union
+
+import logging
+import os
+import torch
+
+GCS_BUCKET = 'datcom-nl-models'
+EMBEDDINGS = 'prod/'
+EMBEDDINGS_FILENAME = "embeddings_us_filtered_2023_1_26_22_15_41.csv"
+TEMP_DIR = '/tmp/'
+MODEL_NAME = 'all-MiniLM-L6-v2'
+
+
+class Embeddings:
+  """Manages the embeddings."""
+
+  def __init__(self) -> None:
+    self.model: SentenceTransformer = SentenceTransformer(MODEL_NAME)
+    self.dataset_embeddings: torch.Tensor = None
+    self._download_embeddings()
+    self.dcids: List[str] = []
+    self.sentences: List[str] = []
+
+    logging.info('Loading embeddings file')
+    try:
+      ds = load_dataset('csv',
+                        data_files=os.path.join(TEMP_DIR,
+                                                f'{EMBEDDINGS_FILENAME}'))
+    except:
+      error_str = "No embedding could be loaded."
+      logging.error(error_str)
+      raise Exception("No embedding could be loaded.")
+
+    df = ds["train"].to_pandas()
+    self.dcids = df['dcid'].values.tolist()
+    df = df.drop('dcid', axis=1)
+    # Also get the sentence mappings.
+    self.sentences = []
+    if 'sentence' in df:
+      self.sentences = df['sentence'].values.tolist()
+      df = df.drop('sentence', axis=1)
+
+    self.dataset_embeddings = torch.from_numpy(df.to_numpy()).to(torch.float)
+
+  def _download_embeddings(self):
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(bucket_name=GCS_BUCKET)
+    blobs = bucket.list_blobs(prefix=EMBEDDINGS)  # Get list of files
+    for blob in blobs:
+      try:
+        _, filename = os.path.split(blob.name)
+        if filename == EMBEDDINGS_FILENAME:
```

Comment:
@shifucun this is where we should be looking for the latest timestamp.

Reply:
Can we only read the embedding file as specified?

Reply:
Yup, done now